### PR TITLE
build: transform dynamic imports to static imports for umd build

### PR DESCRIPTION
**Pull request by ChristianMurphy**
rollup currently does not support dynamic import in iife or umd builds.
this transforms dynamic imports to static imports to allow umd builds to continue to work.

an example of the input and output of the plugin can be seen here https://astexplorer.net/#/gist/16c8cd275ea5b0df3da862d2fb5add9a/8ce0535267e300bc149d78a9b738b5740df63a8b

### CODE REVIEW

File: scripts/babel-plugin-dynamic-to-static-import.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        // find dynamic import expression
+        if (
+          t.isCallExpression(path.node.expression) &&
+          t.isImport(path.node.expression.callee)
+        ) {
+          // create static import
+          const staticImport = t.importDeclaration(
+            [],
+            t.stringLiteral(path.node.expression.arguments[0].value)
+          );
```
Comment (ChristianMurphy) [inline scripts/babel-plugin-dynamic-to-static-import.js:16]: this assumes the dynamic import will be in the form

```js
import("name")
```

which will transform to

```js
import "name";
```

it does **not** handle

```js
import("name").then((name) => /* do something */);
```

which would transform to

```js
import name from "name";
```

**Comment by btopro** [comment]
So is there a reason that `@babel/plugin-syntax-dynamic-import` isn't doing this transform? We do a handful of the Promise based dynamic import methods throughout our stack (very few but I think HAX cheats on performance big time by doing this in its boostrap) so would those still fail then?

File: scripts/babel-plugin-dynamic-to-static-import.js (hunk 1 above)
Comment (ChristianMurphy) [reply scripts/babel-plugin-dynamic-to-static-import.js:18]: resolved in c9f9feb

**Comment by ChristianMurphy** [comment]
> So is there a reason that `@babel/plugin-syntax-dynamic-import` isn't doing this transform?

Good question, `@babel/plugin-syntax-dynamic-import` resolves `'import' and 'export' may only appear at the top level`.
https://astexplorer.net/#/gist/16c8cd275ea5b0df3da862d2fb5add9a/e2df98c501907fc469a9d5659c5be9287f09bd72
but that's it, it parses the new syntax, and stops.

> We do a handful of the Promise based dynamic import methods throughout our stack (very few but I think HAX cheats on performance big time by doing this in its boostrap) so would those still fail then?

c9f9feb should resolve this.
Here is a sample of how the new plugin attempts to resolve these: https://astexplorer.net/#/gist/16c8cd275ea5b0df3da862d2fb5add9a/8ce0535267e300bc149d78a9b738b5740df63a8b

**Comment by btopro** [comment]
ooo ok that makes more sense. I wonder if this isn't something that could be a reusable library that should be published for others since this is beyond weeds we are in :)

**Comment by ChristianMurphy** [comment]
>  I wonder if this isn't something that could be a reusable library that should be published for others since this is beyond weeds we are in

It probably is, I'll look into publishing it as an npm module.

**Comment by ChristianMurphy** [comment]
Gah, looks like I missed one more thing

```
await import(`${this.basePath}${this.list[tag]}`)
```

a dynamic import with a dynamic path can't be translated to a static import.
I'll add a condition to skip over cases like this.

**Comment by btopro** [comment]
I do enjoy generating problems w/ how I use things to the spec but not the standard way compilers are looking for ;)

**Pull request merged**